command/agent/retry_join.go

```diff
@@ -30,6 +44,87 @@ type DiscoverInterface interface {
 	Names() []string
 }
 
+// NetaddrsInterface is an interface for go-netaddrs to ease testing
+type NetaddrsInterface interface {
+	IPAddrs(ctx context.Context, cfg string, l netaddrs.Logger) ([]net.IPAddr, error)
+}
+
+type netAddrs struct{}
+
+func (n *netAddrs) IPAddrs(ctx context.Context, cfg string, l netaddrs.Logger) ([]net.IPAddr, error) {
+	return netaddrs.IPAddrs(ctx, cfg, l)
+}
+
+// autoDiscover uses go-netaddrs and go-discover to discover IP addresses when
+// auto-joining clusters
+//
+// autoDiscover implements AutoDiscoverInterface
+type autoDiscover struct {
+	netAddrs   NetaddrsInterface
+	goDiscover DiscoverInterface
+}
+
+// Addrs looks up and returns IP addresses specified by cfg.
+//
+// If cfg is a IPv4 or IPv6 address, return the stringified IP address.
+//
+// If cfg has a dns= prefix, the configuration is treated as a DNS lookup on the
+// name following the prefix.
+//
+// If cfg has an exec= prefix, IP addresses are looked up by executing the command
+// after exec=. The command may include optional arguments. Command arguments
+// must be space separated (spaces in argument values can not be escaped).
+// The command may output IPv4 or IPv6 addresses, and IPv6 addresses can
+// optionally include a zone index.
+//
+// The executable must follow these rules:
+//
+//	on success - exit 0 and print whitespace delimited IP addresses to stdout.
+//	on failure - exits with a non-zero code, and should print an error message
+//	             of up to 1024 bytes to stderr.
+//
+// If cfg has a provider= prefix, IP addresses are looked up using the go-discover
+// provider specified in cfg.
+func (d autoDiscover) Addrs(cfg string, logger hclog.Logger) (addrs []string, err error) {
+	ip := net.ParseIP(cfg)
+	if ip != nil {
+		return []string{cfg}, nil
+	}
+
+	if !cfgPrefixes.MatchString(cfg) {
+		return nil, fmt.Errorf("auto-discover configurations must begin with: exec=, provider=, or dns=; got %s", cfg)
+	}
+
+	var ipAddrs []net.IPAddr
+	prefixMatches := cfgPrefixes.FindStringSubmatch(cfg)
+	cfgPrefix := prefixMatches[1]
+	cfgValue := prefixMatches[2]
+	switch cfgPrefix {
+	case "exec":
+		ipAddrs, err = d.netAddrs.IPAddrs(context.Background(),
+			cfg,
+			logger)
+
+		for _, addr := range ipAddrs {
+			addrs = append(addrs, addr.String())
+		}
+	case "dns":
+		ipAddrs, err = d.netAddrs.IPAddrs(context.Background(),
+			cfgValue,
+			logger)
+		for _, addr := range ipAddrs {
+			addrs = append(addrs, addr.String())
+		}
+	case "provider":
+		addrs, err = d.goDiscover.Addrs(cfg, logger.StandardLogger(&hclog.StandardLoggerOptions{InferLevels: true}))
+	default:
```

Comment:
Your mention of docs had me look at the existing docs, and I learned we have [existing support for DNS](https://developer.hashicorp.com/nomad/docs/configuration/server_join#via-domains-or-dns). If you follow the `serverJoin` and `clientJoin` functions from where they're injected by the caller in `command.go`, the server join code path resolves DNS names down in the Serf library, and the client join code path resolves DNS names itself.

So that complicates our error handling a bit. Fortunately `exec=` and `provider=` and `dns=` aren't valid domain names because of the `=`, so I think what we can do is try to resolve the address here and only if we can't resolve an address do we return an error. We can probably just remove the `dns=` case as well.

Sorry I missed that in my first review. There's a lot of very old code here. :blush:

Reply:
No worries! I really should have begun my implementation with a read of the docs :)

Reply:
This change greatly simplified `Addrs()` in he latest commit. 

Everything that is not `exec=` or `provider=` prefixed is passed along as-is, which reverts this to the previous behavior. With the `exec=` functionality added, of course. 

All addrs returned by `Addrs()` that are not IPs get resolved by Serf down the line in the server's `Join()`.

Reply:
:+1: